readthedocs/subscriptions/products.py

```diff
@@ -0,0 +1,113 @@
+"""
+Read the Docs products.
+
+This module contains the dataclasses that represent the products available.
+The available products are defined in the ``RTD_PRODUCTS`` setting.
+"""
+from dataclasses import dataclass
+
+from django.conf import settings
+
+from readthedocs.subscriptions.constants import FEATURE_TYPES
+
+
+@dataclass(slots=True)
+class RTDProductFeature:
+    # readthedocs.subscriptions.constants.TYPE_*
+    type: str
+    # Depending on the feature type, this number can represent a numeric limit, number of days, etc.
+    value: int = 0
+    # If this feature is unlimited for this product.
+    unlimited: bool = False
+    description: str = ""
+
+    def get_description(self):
+        return self.description or dict(FEATURE_TYPES).get(self.type)
+
+    def to_item(self):
+        """
+        Return a tuple with the feature type and the feature itself.
+
+        Useful to use it as a dictionary item.
+        """
+        return self.type, self
+
+
+@dataclass(slots=True)
+class RTDProduct:
+
+    """A local representation of a Stripe product."""
+
+    stripe_id: str
+    features: dict[str, RTDProductFeature]
+    # If this product should be available to users to purchase.
+    listed: bool = False
+
+    def to_item(self):
+        """
+        Return a tuple with the stripe_id and the product itself.
+
+        Useful to use it as a dictionary item.
+        """
+        return self.stripe_id, self
+
+
+def get_product(stripe_id) -> RTDProduct:
+    """Return the product with the given stripe_id."""
+    return settings.RTD_PRODUCTS.get(stripe_id)
+
+
+def get_listed_products():
+    """Return a list of products that are available to users to purchase."""
+    return [product for product in settings.RTD_PRODUCTS.values() if product.listed]
+
+
+def get_products_with_feature(feature_type):
+    """Return a list of products that have the given feature."""
+    return [
+        product
+        for product in settings.RTD_PRODUCTS.values()
+        if feature_type in product.features
+    ]
+
+
+def get_feature(obj, feature_type) -> RTDProductFeature:
```

Comment:
I'm starting to use some type hints here, not strictly required here, but they do make the LSP more useful during development.

Reply:
I was going through all the calls to `get_feature` and they seem to all be aware of weather `obj` is a `Project` or `Organization`? Perhaps better to make a call signature like `get_feature(feature_type, project:Project=None, organization:Organization=None)` ?